File: DSA/Ansh-Gulati/Day-10/Day-10_Question-2.cpp
#include <iostream>
#include <stack>
using namespace std;

class Queue
{
private:
    stack<int> stack1;
    stack<int> stack2;

    void moveStack1ToStack2()
    {
        while (!stack1.empty())
        {
            stack2.push(stack1.top());
            stack1.pop();
        }
    }

public:
    void enqueue(int data)
    {
        stack1.push(data);
    }

    int dequeue()
    {
        if (stack2.empty())
        {
            if (stack1.empty())
            {
                throw out_of_range("Queue is empty");
            }
            moveStack1ToStack2();
        }
        int topVal = stack2.top();
        stack2.pop();
        return topVal;
    }

    void printQueue()
    {
        if (stack2.empty())
        {
            moveStack1ToStack2();
        }
        stack<int> temp = stack2;
        while (!temp.empty())
        {
            cout << temp.top() << " ";
            temp.pop();
        }
        cout << endl;
    }
};

int main()
{
    Queue q;

    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    q.enqueue(4);
    q.enqueue(5);

    cout << "Queue: ";
    q.printQueue();

    cout << "Dequeued: " << q.dequeue() << endl;
    cout << "Dequeued: " << q.dequeue() << endl;
    cout << "Dequeued: " << q.dequeue() << endl;

    cout << "Queue: ";
    q.printQueue();

    return 0;
}

File: DSA/Manav-Rajput/Day10_question_2.java
import java.util.Stack;

public class QueueUsingStacks {

    private Stack<Integer> stack1;
    private Stack<Integer> stack2;

    //constructor to initialize the two stacks
    public QueueUsingStacks() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }
    //method to add an element to the queue (enqueue)
    public void enqueue(int data) {
        stack1.push(data);
    }
    //method to remove and return an element from the queue (dequeue)
    public int dequeue() {
        //if both stacks are empty, throw an exception
        if (stack1.isEmpty() && stack2.isEmpty()) {
            throw new RuntimeException("Queue is empty");
        }
        //if stack2 is empty transfer all elements from stack1 to stack2
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.push(stack1.pop());
            }
        }
        //return the top element from stack2
        return stack2.pop();
    }

    public static void main(String[] args) {
        //create a queue using two stacks
        QueueUsingStacks queue = new QueueUsingStacks();
        //enqueue elements into the queue
        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);
        //dequeue elements from the queue and print them
        System.out.println(queue.dequeue()); // Output: 1
        System.out.println(queue.dequeue()); // Output: 2
        System.out.println(queue.dequeue()); // Output: 3
    }
}
